packages/@aws-cdk/aws-apigateway/lib/deployment.ts

```diff
@@ -0,0 +1,160 @@
+
+    constructor(parent: cdk.Construct, id: string, props: DeploymentProps) {
+        super(parent, id);
+
+        this.resource = new LatestDeploymentResource(this, 'Resource', {
+            description: props.description,
+            restApiId: props.api.restApiId,
+        });
+
+        if (props.retainDeployments === undefined || props.retainDeployments) {
+            this.resource.options.deletionPolicy = cdk.DeletionPolicy.Retain;
+        }
+
+        this.api = props.api;
+        this.deploymentId = new DeploymentId(() => this.resource.ref);
+    }
+
+    /**
+     * Adds a dependency for this deployment. Should be called by all resources and methods
+     * so they are provisioned before this Deployment.
+     */
+    public addDependency(dep: cdk.IDependable) {
+        this.resource.addDependency(dep);
+    }
+
+    /**
+     * Adds a component to the hash that determines this Deployment resource's
+     * logical ID.
+     *
+     * This should be called by constructs of the API Gateway model that want to
+     * invalidate the deployment when their settings change. The component will
+     * be resolve()ed during synthesis so tokens are welcome.
+     */
+    public addToLogicalId(data: any) {
+        this.resource.addToLogicalId(data);
+    }
+}
+
+class LatestDeploymentResource extends cloudformation.DeploymentResource {
+    private originalLogicalId: string;
+    private customLogicalId?: string;
+    private hashComponents = new Array<any>();
+
+    constructor(parent: cdk.Construct, id: string, props: cloudformation.DeploymentResourceProps) {
+        super(parent, id, props);
+
+        this.originalLogicalId = this.logicalId;
+
+        Object.defineProperties(this, {
+            logicalId: {
+                get: () => {
+                    if (!this.customLogicalId) {
+                        throw new Error('The logical ID of this resource cannot be evaluated eagerly. Use: new cdk.Token(() => foo.logicalId)');
+                    }
+                    return this.customLogicalId;
+                }
+            },
+            ref: {
+                get: () => new cdk.Token(() => ({ Ref: this.customLogicalId }))
+            },
+        });
+    }
+
+    /**
+     * Allows adding arbitrary data to the hashed logical ID of this deployment.
+     * This can be used to couple the deployment to the API Gateway model.
+     */
+    public addToLogicalId(data: unknown) {
+
+        // if the construct is locked, it means we are already synthesizing and then
+        // we can't modify the hash because we might have already calculated it.
+        if (this.locked) {
+            throw new Error('Cannot modify the logical ID when the construct is locked');
+        }
+
+        this.hashComponents.push(data);
+    }
+
+    /**
+     * Hooks into synthesis to calculate a logical ID that hashes all the components
+     * add via `addToLogicalId`.
+     */
+    public validate() {
```

Comment:
Yeah, it's hack. Yeah, we should generalize. I wanted to do this as part of #282, but maybe I should implement a simple overload for now like you suggest

Reply:
I'll punt this addition to the core to a subsequent commit. It's related to the change I am working on for supporting synthesis at the construct-level, and then it will be possible to define those hooks properly.